### PR TITLE
Make IDA work with new sundials.

### PR DESCRIPTION
**Pull request by luca-heltai**
I'd like to have this merged in 9.3.

IDA is the only missing piece of sundials that was still not working with version > 4.0.0.

This PR <s>does not</s> makes use of all the new facilities of SUNDIALS <s>but at least makes it run with the new version</s> (thanks to @bangerth for his work on KINSOL, which @nicola-giuliani copied almost verbatim).

<s>@sebproell , this is a temporary patch. I'd like to incorporate your changes, if time permits.</s>

### CODE REVIEW

File: include/deal.II/sundials/ida.h [hunk 1]
```diff
@@ -707,6 +707,15 @@ namespace SUNDIALS
     std::function<int(const VectorType &rhs, VectorType &dst)>
       solve_jacobian_system;
 
+#  if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
+    /**
+     * A wrapper for a SUNLinearSolver.
+     */
+    LinearSolveFunction<VectorType> solve_jacobian_system_sundials;
+    std::unique_ptr<internal::LinearSolverWrapper<VectorType>>
+      jacobian_linear_solver;
```
Comment (bangerth) [inline include/deal.II/sundials/ida.h:716]: This is in the public interface. Can you explain in more detail what this function is supposed to do? It is also referencing an `internal::` class, which looks awkward in the public interface...

File: source/sundials/ida.cc [hunk 2]
```diff
@@ -151,11 +159,79 @@ namespace SUNDIALS
...
+                         N_Vector /* ignored */,
+                         SUNMatrix /* ignored */,
+                         void *user_data,
+                         N_Vector /* ignored */,
+                         N_Vector /* ignored */,
+                         N_Vector /* ignored */)
+    {
+      Assert(user_data != nullptr, ExcInternalError());
+      IDA<VectorType> &solver = *static_cast<IDA<VectorType> *>(user_data);
+      std::cout << "setting jacobian" << std::endl;
```
Comment (bangerth) [inline source/sundials/ida.cc:181]: please remove

File: source/sundials/ida.cc [hunk 3]
```diff
@@ -151,11 +159,79 @@ namespace SUNDIALS
...
+    t_dae_jacobian_setup(realtype tt,
+                         realtype cj,
+                         N_Vector yy,
+                         N_Vector yp,
+                         N_Vector /* ignored */,
+                         SUNMatrix /* ignored */,
+                         void *user_data,
+                         N_Vector /* ignored */,
+                         N_Vector /* ignored */,
+                         N_Vector /* ignored */)
```
Comment (bangerth) [inline source/sundials/ida.cc:177]: Out of curiosity, what are all of these arguments? In the case of KINSOL, I ended up adding a new function object that actually makes use of these vectors because they turned out to be useful.

File: source/sundials/ida.cc [hunk 4]
```diff
@@ -151,11 +159,79 @@ namespace SUNDIALS
...
+
+
+
+    template <typename VectorType>
+    int
+    t_dae_solve(SUNLinearSolver LS,
+                SUNMatrix /*ignored*/,
+                N_Vector x,
+                N_Vector b,
+                realtype /*tol*/)
```
Comment (bangerth) [inline source/sundials/ida.cc:208]: same here with the tolerance. You can probably just copy the same structure from my last KINSOL patch (#12031).

**Review by bangerth — CHANGES_REQUESTED** [review]

File: source/sundials/ida.cc (hunk 3 above)
Comment (nicola-giuliani) [reply source/sundials/ida.cc]: These are actually three tmp vectors that IDA provides in case they are needed. For the time being they are not, and I have no idea what they could be, but I stated what they are in the comment.

File: source/sundials/ida.cc (hunk 4 above)
Comment (nicola-giuliani) [reply source/sundials/ida.cc]: I tried to copy your structure. The only difference is that I check if the user has implemented the new version or the deprecated one. If non of the two is implemented a throw an assertion

File: include/deal.II/sundials/ida.h (hunk 1 above)
Comment (nicola-giuliani) [reply include/deal.II/sundials/ida.h]: That was a leftover...sorry ;).

File: source/sundials/ida.cc (hunk 2 above)
Comment (nicola-giuliani) [reply source/sundials/ida.cc]: again a leftover, again sorry

**Comment by nicola-giuliani** [comment]
@bangerth, @luca-heltai  this patch only covers the basic interface that was already provided in deal.II for SUNDIALS <= 4. My major doubt concerns the resid function that should be provided to the SUNLinSolver. For the time being it simply returns a nullptr so I presume IDA does not use it. It should be an internal check if the preconditioned residual is also a solution but I don't know how to use it, with @luca-heltai we tried to understand it unsuccessfully.

**Comment by luca-heltai** [comment]
This is not working yet. We must have messed up with a few `ifdefs`. I'm working on it now.

**Comment by luca-heltai** [comment]
@sebproell , I used the new interface in the last commit. Would you mind checking the parts that I've added to make sure things are correct? I had to had one more interface to `n_vector.templates.h` that IDA uses. I'll now coordinate with Nicola, in order to create a few more tests and verify that the new interface works for all corner cases. Admittedly, ida is not well tested, and we should make sure that all things work.

**Comment by luca-heltai** [comment]
I'm not sure how to proceed with the `solve_jacobian_system` function. In version < 4.0.0, we provided no option to pass a tolerance. When using version > 4.0.0, this becomes the recommended behavior, but we allow also not to specify it. Should we guard the deprecation, i.e., add `ifdefs` around the member `solve_jacobian_system`, and definie it twice, one without deprecation for version < 4.0.0, and one with deprecation for version > 4.0.0?

File: include/deal.II/sundials/ida.h [hunk 5]
```diff
@@ -581,13 +595,25 @@ namespace SUNDIALS
...
+    set_n_iter(const int n_iter)
+    {
+      this->n_iter = n_iter;
+    }
+
+    int
+    get_n_iter() const
+    {
+      return this->n_iter;
+    }
```
Comment (peterrum) [inline include/deal.II/sundials/ida.h:616]: The documentation is missing here. Why not make unsigned int?

File: include/deal.II/sundials/n_vector.templates.h [hunk 6]
```diff
@@ -635,6 +639,24 @@ SUNDIALS::internal::NVectorOperations::weighted_rms_norm(N_Vector x, N_Vector w)
 
 
 
+template <typename VectorType>
+realtype
+SUNDIALS::internal::NVectorOperations::weighted_rms_norm_mask(N_Vector x,
+                                                              N_Vector w,
+                                                              N_Vector mask)
+{
+  // TODO copy can be avoided by a custom kernel
```
Comment (peterrum) [inline include/deal.II/sundials/n_vector.templates.h:648]: I guess this is OK since some of the "old" functions are also doing this like this.

**Review by peterrum — APPROVED** [review]
Thanks @nicola-giuliani  and @luca-heltai! Looks good.

I would suggest that we try to merge this PR tomorrow so that it is part of the first release candidate!

File: include/deal.II/sundials/ida.h (hunk 5 above)
Comment (luca-heltai) [reply include/deal.II/sundials/ida.h:616]: It is needed by sundials as an int.

File: include/deal.II/sundials/n_vector.templates.h [hunk 7]
```diff
@@ -635,6 +639,24 @@ SUNDIALS::internal::NVectorOperations::weighted_rms_norm(N_Vector x, N_Vector w)
...
+  VectorType tmp         = *unwrap_nvector_const<VectorType>(x);
+  auto *     w_dealii    = unwrap_nvector_const<VectorType>(w);
+  auto *     mask_dealii = unwrap_nvector_const<VectorType>(mask);
+  const auto n           = tmp.size();
+  tmp.scale(*w_dealii);
+  tmp.scale(*mask_dealii);
+  return tmp.l2_norm() / std::sqrt(n);
```
Comment (sebproell) [inline include/deal.II/sundials/n_vector.templates.h:655]: This fails for an empty vector x.

**Review by sebproell — COMMENTED** [review]
Would you mind adding tests for the new N_Vector operation?

**Comment by sebproell** [comment]
@luca-heltai Nice to see the N_Vector implementation put to use in this PR! I only checked wrt to these changes as I don't know anything about IDA. Your usage of it is as I intended 👍

File: include/deal.II/sundials/n_vector.templates.h (hunk 7 above)
Comment (luca-heltai) [reply include/deal.II/sundials/n_vector.templates.h:655]: Just like the other one. Sundials should throw in that case.

**Review by peterrum — APPROVED** [review]
If there are no objections, I will merge this PR in the evening (CET).

**Comment by nicola-giuliani** [comment]
@peterrum I'm having troubles running the interface with sundials 5.7.0. on my machine. @luca-heltai run the tests successfully with SUNDIALS 5.6.1. I am recompiling deal to check what's happening. Apparently SUNDIALS claims that a vector operation is not implemented when it initialises the solver.

**Comment by peterrum** [comment]
> Everything works fine with SUNDIALS 5.7.0. I had a problem of the installation of deal but a fresh compilation addressed the issue.

Yes, I can verify this! Thanks.

**Comment by peterrum** [comment]
@nicola-giuliani What about the other IDA versions? If they are working, let' rebase this branch and remove the merge conflict.

**Comment by nicola-giuliani** [comment]
@peterrum rebase done, @luca-heltai should have implemented just one new function for the n vectors.

**Review by peterrum — APPROVED** [review]

**Comment by masterleinad** [comment]
```
sundials/kinsol_01.debug: RUN failed. ------ Additional output on stdout/stderr:
[KINDLS ERROR]  KINDlsSetLinearSolver
  Both LS and A must be non-NULL
--------------------------------------------------------
An error occurred in line <494> of file </jenkins/workspace/dealii_PR-12200/source/sundials/kinsol.cc> in function
    unsigned int dealii::SUNDIALS::KINSOL<VectorType>::solve(VectorType&) [with VectorType = dealii::Vector<double>]
The violated condition was: 
    status >= 0
Additional information: 
    One of the SUNDIALS KINSOL internal functions returned a negative
    error code: -3. Please consult SUNDIALS manual.
--------------------------------------------------------
```

**Comment by luca-heltai** [comment]
This has nothing to do with IDA, but with Kinsol. See #12223

File: include/deal.II/sundials/ida.h [hunk 8]
```diff
@@ -703,10 +738,61 @@ namespace SUNDIALS
...
+     * - >0: Recoverable error (IDAReinit will be called if this happens, and
+     *       then last function will be attempted again
+     * - <0: Unrecoverable error the computation will be aborted and an
+     * assertion will be thrown.
+     */
+    std::function<int(const VectorType &rhs,
+                      VectorType &      dst,
+                      int &             n_iter,
+                      const double      tolerance)>
+      solve_jacobian_system_up_to_tolerance;
```
Comment (bangerth) [inline include/deal.II/sundials/ida.h:794]: @luca-heltai I'm coming late to the game here, but would it be possible to use the same function name as for the KINSOL interface? There, I used `solve_with_jacobian`. I think it would be nice if we could use consistent names...

**Comment by luca-heltai** [comment]
I'm not sure that 
- solve_with_jacobian
is any different/better/self-documenting than
- solve_jacobian_system

I had to look it up several times, and I was getting very confused until I understood what you wanted to do. How about we rename the KINSOL ones as `solve_jacobian_sytem_up_to_tolerance`, since this is what that function does?

**Comment by bangerth** [comment]
Discussion continued in #12267.
